### PR TITLE
Pt 159406473 sc continuous chain monitoring 4 -- review

### PR DESCRIPTION
**Pull request by lucafavatella**
TODO:
* [x] Ensure CI tests pass.

### CODE REVIEW

File: apps/aechannel/src/aesc_fsm.erl [hunk 1]
```diff
@@ -555,23 +555,24 @@ dry_run_contract(Fsm, #{contract    := _,
 %% FSM initialization
 
 init(#{opts := Opts0} = Arg) ->
-    #{role := Role, client := Client} = Opts0,
+    {Role, Opts1} = maps:take(role, Opts0),
```
Comment (lucafavatella) [inline apps/aechannel/src/aesc_fsm.erl]: Integration test [fails on CI](https://circleci.com/gh/aeternity/aeternity/38675#artifacts/containers/0):
```
%%% aehttp_integration_SUITE ==> all.channel_websocket.sc_ws_leave_reestablish: FAILED
%%% aehttp_integration_SUITE ==> {{badmatch,{error,rejected}},
 [{aehttp_integration_SUITE,sc_ws_reestablish_,2,
                            [{file,"/home/builder/aeternity/apps/aehttp/test/aehttp_integration_SUITE.erl"},
                             {line,3903}]},
  {aehttp_integration_SUITE,sc_ws_leave_reestablish,1,
                            [{file,"/home/builder/aeternity/apps/aehttp/test/aehttp_integration_SUITE.erl"},
                             {line,5421}]},
```

From [node log](https://38675-99802036-gh.circle-artifacts.com/0/home/builder/aeternity/_build/test/logs/ct_run.aeternity_ct%40localhost.2019-05-30_16.24.24/lib.aehttp.aehttp_integration_SUITE.logs/run.2019-05-30_16.28.54/log_private/aehttp_integration_SUITE/dev1/log/aeternity.log):
```
2019-05-30 16:31:39.448 [error] <0.273.0> Lager event handler error_logger_lager_h exited with reason {'EXIT',{{case_clause,[<0.3433.0>,undefined,error,function_clause,undefined,[{aesc_offchain_state,my_pubkey,[#{channel_reserve => 2,client => <0.3431.0>,existing_channel_id => <<175,222,4,4,97,99,94,149,27,126,99,176,191,51,38,227,16,31,202,152,104,214,19,142,9,191,131,1,27,129,143,32>>,initiator => <<177,181,119,188,211,39,203,57,229,94,108,2,107,214,167,74,27,53,222,108,6,80,196,174,81,239,171,117,158,65,91,102>>,initiator_amount => 70000000000000,lock_period => 10,log_keep => 10,minimum_depth => ...,...}],...},...]]},...}}
```

File: apps/aechannel/src/aesc_fsm.erl [hunk 2]
```diff
@@ -217,7 +249,10 @@ record_fields(Other) -> aesc_offchain_state:record_fields(Other).
 -define(LOG_CAUGHT(Err), lager:debug("CAUGHT ~p", [Err])).
 -endif.
 
-callback_mode() -> [state_functions, state_enter].
+callback_mode() ->
+    [ state_functions %% "... Module:StateName/3, is used."
+    , state_enter %% "... at every **state change**, call the state callback with arguments `(enter, OldState, Data)`. ... a state enter call will be done right before entering the initial state ..."
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]:
```suggestion
    , state_enter     %% "... at every **state change**, call the state callback
                      %% with arguments `(enter, OldState, Data)`. ... a state
                      %% enter call will be done right before entering the initial
                      %% state ..."
```

File: apps/aechannel/src/aesc_fsm.erl [hunk 3]
```diff
@@ -552,25 +586,33 @@ dry_run_contract(Fsm, #{contract    := _,
 %% ======================================================================
 %% FSM initialization
 
+-spec init(term()) -> {ok, InitialState, data(), [{timeout, Time::pos_integer(), InitialState}, ...]} when InitialState :: state_name().
```
Comment (uwiger) [inline apps/aechannel/src/aesc_fsm.erl]:
```suggestion
-spec init(term()) -> {ok, InitialState, data(), [{timeout, Time::pos_integer(),
                                                   InitialState}, ...]}
                          when InitialState :: state_name().
```

Comment (uwiger) [reply apps/aechannel/src/aesc_fsm.erl]: (only adding line breaks + indentation to keep within ca 80 char width)

File: apps/aechannel/src/aesc_fsm.erl (hunk 2 above)
Comment (uwiger) [reply apps/aechannel/src/aesc_fsm.erl]: (adding line breaks + indentation to keep within ca 80 chars width)

File: apps/aecore/src/aec_db.erl [hunk 4]
```diff
@@ -118,10 +118,7 @@
         {Record, tab(Mode, Record, record_info(fields, Record), Extra)}).
 
 %% start a transaction if there isn't already one
--define(t(Expr), case get(mnesia_activity_state) of undefined ->
-                         transaction(fun() -> Expr end);
-                     _ -> Expr
-                 end).
+-define(t(Expr), ensure_transaction(fun() -> Expr end)).
```
Comment (uwiger) [inline apps/aecore/src/aec_db.erl]: This change actually is undesireable. The `aec_db` functions should simply execute in the existing context, creating a transaction context if there is none.

**Pull request merged**